Marlin/src/feature/runout.h

```diff
@@ -153,155 +158,170 @@ class FilamentSensorTypeBase {
-
-      // Clear motion triggers for next block
-      motion_detected = 0;
-    }
 
-    static inline void run() { poll_motion_sensor(); }
-};
+#endif // !FILAMENT_MOTION_SENSOR
 
 /********************************* RESPONSE TYPE *********************************/
 
 #if FILAMENT_RUNOUT_DISTANCE_MM > 0
 
-  // The RunoutResponseDelayed will trigger an runout event only after
-  // RUNOUT_DISTANCE_MM of filament have been fed after a runout condition.
+  // RunoutResponseDelayed triggers a runout event only if the length
+  // of filament specified by FILAMENT_RUNOUT_DISTANCE_MM has been fed
+  // during a runout condition.
   class RunoutResponseDelayed {
     private:
-      static int32_t steps_since_detection[EXTRUDERS];
-
-      static float get_mm_since_runout(const uint8_t extruder) {
-        return (steps_since_detection[extruder] / planner.settings.axis_steps_per_mm[E_AXIS_N(extruder)]);
-      }
+      static float runout_mm_countdown[EXTRUDERS];
 
     public:
-      static float runout_distance_mm;
-
-      static inline bool has_runout() {
-        return get_mm_since_runout(active_extruder) > runout_distance_mm;
-      }
+      static float runout_threshold_mm;
 
-      static inline void filament_present(const uint8_t extruder) {
-        steps_since_detection[extruder] = 0;
+      static void reset() {
+        LOOP_L_N(i, EXTRUDERS) filament_present(i);
       }
 
       static inline void run() {
         #ifdef FILAMENT_RUNOUT_SENSOR_DEBUG
-          static uint16_t r = 0;
-          if ((r++ % 24000) == 0) {
-            SERIAL_PROTOCOLPGM("mm since filament detection: ");
-            LOOP_L_N(i, NUM_RUNOUT_SENSORS) {
-              if (i > 0) SERIAL_PROTOCOLPGM(", ");
-              SERIAL_PROTOCOL(get_mm_since_runout(i));
+          static millis_t t = 0;
+          const millis_t ms = millis();
+          if (ELAPSED(ms, t)) {
+            t = millis() + 1000UL;
+            LOOP_L_N(i, EXTRUDERS) {
+              serialprintPGM(i ? PSTR(", ") : PSTR("Remaining mm: "));
+              SERIAL_PROTOCOL(runout_mm_countdown[i]);
             }
             SERIAL_EOL();
           }
         #endif
       }
 
-      static void reset() {
-        LOOP_L_N(i, NUM_RUNOUT_SENSORS) steps_since_detection[i] = 0;
+      static inline bool has_run_out() {
+        return runout_mm_countdown[active_extruder] < 0;
+      }
+
+      static inline void filament_present(const uint8_t extruder) {
+        runout_mm_countdown[extruder] = runout_threshold_mm;
       }
 
-      static inline void block_complete(const block_t *b) {
-        steps_since_detection[b->extruder] += TEST(b->direction_bits, E_AXIS) ? -b->steps[E_AXIS] : b->steps[E_AXIS];
+      static inline void block_complete(const block_t* const b) {
+        const uint8_t e = b->extruder;
+        const int32_t steps = b->steps[E_AXIS];
+        runout_mm_countdown[e] -= (TEST(b->direction_bits, E_AXIS) ? -steps : steps) * planner.steps_to_mm[E_AXIS_N(e)];
```

Comment:
This does happen less often than `run()`, but since it happens in the stepper ISR it's an open question. At least it now has the benefit of using multiplication rather than division.

Reply:
Why not just countdown steps from the pre-calculated initial value? I.e. you could avoid the multiplication in `block_complete` if you count down steps instead of mm. The initial amount of steps ( for each extruder )  would be the ` initial_mm * planner.steps_to_mm[E_AXIS_N( for each extruder )]`

Reply:
ah, ok. so disregard please. Wishful thinking on my part.

Reply:
True. You could precalculate the values as 32-bits per extruder, but storage is very much at a permium on this AVR microcontrollers, so I don't think that would fly :)